core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -525,6 +525,16 @@ class LogSegment private[log] (val log: FileRecords,
     }
   }
 
+  /**
+    * @return the first batch timestamp if the timestamp is available and has been previously loaded. Otherwise return 0
+    */
+  def getFirstBatchTimestamp() : Long = {
```

Comment:
Nit: timestamp won't be negative in this scenario right ? if it is non-negative value, we can change it to
`rollingBasedTimestamp.getOrElse(0L)`

Reply:
In timeWaitedForRoll, we don't make assumption that timestamp is non-negative, so we cannot make assumption here. the timestamp is populated from the message record.